### PR TITLE
Restructure our cext patching code

### PR DESCRIPTION
**Pull request by aardvark179**
This PR doesn't change the actual patching process, but it does separate the patches out per gem to help reduce conflicting changes. It also changes the format of the hashes that specify the patches, and adds a method to turn those into an internal representation.

This change should allow us to add more file matching options - I'd like to add globbing and better path matching - without changing the structure we specify patches in.

### CODE REVIEW

File: lib/cext/nokogiri_patches.rb [hunk 1]
```diff
@@ -0,0 +1,342 @@
+# Patches nokogiri 1.8.1
```
Comment (nirvdrum) [inline lib/cext/nokogiri_patches.rb]: These patches should work for both 1.8.x and 2.0.x. If you think we should split them, we can.

**Comment by eregon** [comment]
I think the patch files should not be mixed with everything else under lib/cext, so I suggest
`lib/cext/byebug_patches.rb` =>`lib/cext/preprocess/byebug_patches.rb`
or
`lib/cext/byebug_patches.rb` =>`lib/cext/patches/byebug_patches.rb`

File: src/main/c/Makefile [hunk 2]
```diff
@@ -16,7 +16,7 @@ RUBY_LIBRARY := $(ROOT)/lib/cext/ruby.su
 RUBY_HEADERS := $(wildcard $(ROOT)/lib/cext/*.h) $(wildcard $(ROOT)/lib/cext/ruby/*.h)
 RBCONFIG := $(ROOT)/lib/truffle/rbconfig.rb
 RBCONFIG_MKMF := $(ROOT)/lib/truffle/rbconfig-for-mkmf.rb
-PREPROCESS := $(ROOT)/lib/cext/preprocess.rb
+PREPROCESS := $(ROOT)/lib/cext/*.rb
```
Comment (eregon) [inline src/main/c/Makefile]: Do we need `wildcard` like above to be more portable/consistent?

**Comment by nirvdrum** [comment]
Very nice clean-up. I agree with Benoit that the patches should be in their own directory, however.

File: lib/cext/preprocess.rb [hunk 3]
```diff
@@ -6,935 +6,120 @@
...
+    puts "#line 1 \"#{ARGF.filename}\""
 
-  contents = patch(ARGF.filename, File.read(ARGF.filename), Dir.pwd)
-  contents.each_line do |line|
-    puts preprocess(line)
+    contents = patch(ARGF.filename, File.read(ARGF.filename), Dir.pwd)
+    contents.each_line do |line|
+      puts preprocess(line)
+    end
   end
```
Comment (eregon) [inline lib/cext/preprocess.rb]: This "if main" should be moved outside the class, could you do it in another PR since this one is merged?
